[Woo POS] [Barcodes] Play error sound when a barcode scan fails

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Added a functionality to play a sound when barcode scanning fails. Using the same sound as on Android:
- Created a POS sound player to play POS-related sounds
- Added a sound `.mp3` into a project
- Updated tests 

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Enable `pointOfSaleBarcodeScanningi1`
2. Open POS
3. Scan a wrong barcode
4. Confirm that the sound plays

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad Air M2 18.3 Device

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/98ca79ec-886a-4b12-aaa6-577babefb2cc

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Create PointOfSaleSoundPlayer to play sounds on POS

staskus committed:
Play barcode scan failure sound

staskus committed:
Add a test for barcode scan failure sound

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/246">22.6</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on WooCommerce/WooCommerceTests/POS/Models/PointOfSaleAggregateModelTests.swift:
```diff
@@ -24,7 +24,8 @@ struct PointOfSaleAggregateModelTests {
                                                 searchHistoryService: MockPOSSearchHistoryService(),
                                                 popularPurchasableItemsController: MockPointOfSaleItemsController(),
-                                                barcodeScanService: MockPointOfSaleBarcodeScanService())
+                                                barcodeScanService: MockPointOfSaleBarcodeScanService(),
+                                                soundPlayer: MockPointOfSaleSoundPlayer())
```
Keeping iOS16 support is so annoying with this...

joshheald commented on WooCommerce/Classes/POS/Utils/Audio/PointOfSaleSoundPlayer.swift:
```diff
@@ -0,0 +1,33 @@
+            DDLogError("Sound file not found: \(sound.name).\(sound.type)")
+            return
+        }
+        do {
```
I tried with this suggestion... but it doesn't really seem to make any difference to how it sounds with lots of rapid failures.

The leak comment is worth thinking about, but I don't really see how it would leak.

It probably doesn't hurt to add this, but I'll leave it up to you @staskus

staskus replied:
I think a little bit of optimization makes sense here, especially given that if an item is scanned once, it's likely to be scanned multiple times. 

Plus, in an extreme case when we scan multiple wrong items in a row fast, it doesn't make sense to stop the previous sound and start another.

joshheald committed:
Add credit for sound

joshheald reviewed (approved):
Works well. You could integrate the copilot suggestion if you want.

I've added a credit to the licenses section for the sound.

joshheald committed:
Update woo purple in licenses screen

joshheald committed:
Merge branch 'trunk' into woomob-579-woo-posbarcodes-play-error-sound-when-a-barcode-scan-fails

staskus committed:
Optimize PointOfSaleSoundPlayer

- Use a player cache to reuse a player between the plays
- Don't restart the sound it the previous one plays
- Use MainActor for safety when accessing player's cache

joshheald commented:
@staskus I noticed today that we noticably block the main thread when we play the sound, especially the very first time. Perhaps there's a way we can pre-load the sound on another thread to avoid that?

Repro:
1. Type an unknown barcode in the simulator but don't press the button
2. Start tapping other items as fast as you can
3. Tap `Scan`
4. Observe that when the error sound plays it stutters in adding items.

It's also noticable with a stop in the shimmer effect if you're not tapping items and just watching the screen.

staskus commented:
@joshheald thanks for reporting, I'll check
